### PR TITLE
Cherry pick heap-based peer selection into master

### PR DESCRIPTION
**Pull request by abhinav**
We want to release these changes without most of the breaking changes in the `mash` release branch. @junchaowu, can you take a close look at this since you're the primary author of this change? Especially, make sure I didn't miss anything in the cherry-pick.

CC @blampe @breerly

### CODE REVIEW

File: tchannel/tornado/connection.py [hunk 1]
```diff
@@ -122,7 +122,13 @@ def __init__(self, connection, tchannel=None, direction=None):
...
 
         # Map from message ID to futures for responses of outgoing calls.
-        self._outstanding = {}
+        self._outbound_pending_call = {}
+
+        # Map from message ID to outgoing request that is being sent.
+        self._outbound_pending_req = {}
+
+        # Map from message ID to outgoing response that is being sent.
+        self._outbound_pending_res = {}
```
Comment (abhinav) [inline tchannel/tornado/connection.py]: It appears that outbound pending req and res don't actually need references to the actual request and response objects. They can be sets of request IDs, or even just integers maintaining count on the number of ongoing requests and responses. @junchaowu can you confirm this?

Comment (junchaowu) [inline tchannel/tornado/connection.py]: I can say this is true in the future. But for now yes, it only needs the message ID to be stored as a tag.

File: benchmarks/test_concurrent_requests.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+from tornado import gen, ioloop
+from tchannel import TChannel
+
+
+servers = None
+client = None
+
+
+def setUpServer(num):
```
Comment (blampe) [inline benchmarks/test_concurrent_requests.py]: Snake case :)

File: benchmarks/test_concurrent_requests.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        def hello(request):
+            return 'hello'
+        server.listen()
+        servers.append(server)
+     
+    return servers
+
+
+@gen.coroutine
+def setUpClient(servers):
```
Comment (blampe) [inline benchmarks/test_concurrent_requests.py]: Snake case

File: benchmarks/test_concurrent_requests.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            body='hi',
+            hostport=client.hostport,
+        )
+
+    raise gen.Return(client)
+
+
+@gen.coroutine
+def peer_test():
+    global servers, client
```
Comment (blampe) [inline benchmarks/test_concurrent_requests.py]: You don't use servers here. Also why the globals?

**Comment by blampe** [comment]
Looks ok to me.

**Comment by junchaowu** [comment]
it looks okay to me now.

**Pull request merged**